### PR TITLE
feature: add cancellation token to variety of ...WaitFor... APIs

### PR DESCRIPTION
**Pull request by maw136**
fixes: https://github.com/microsoft/playwright/issues/24083
adds CancellationToken where feasible and extends Waiter class to provide CancellationToken as a source to cancel wait

### CODE REVIEW

File: src/Playwright/Core/Waiter.cs [hunk 1]
```diff
@@ -117,6 +117,12 @@ internal void RejectImmediately(Exception exception)
         _immediateError = exception;
     }
 
+    internal void RejectOnCancellation(CancellationToken cancellationToken)
+    {
+        var cancelledTask = Task.FromCanceled(cancellationToken);
+        RejectOn(cancelledTask, () => cancelledTask.Dispose());
```
Comment (campersau) [inline src/Playwright/Core/Waiter.cs:123]: This should probably look something like this:

```suggestion
        if (!cancellationToken.CanBeCanceled) return;
        var cts = new TaskCompletionSource<bool>();
        var registration = cancellationToken.Register(state => cts.SetCanceled((CancellationToken)state), cancellationToken);
        RejectOn(cts.Task, () => registration.Dispose());
```

Comment (maw136) [reply src/Playwright/Core/Waiter.cs:123]: Thank you very much. I was fighting with this part. I found some approaches but nothing worked properly.
I am adding tests to cover cancellation scenarios.

File: src/Playwright/Core/Waiter.cs [hunk 2]
```diff
@@ -218,6 +230,13 @@ internal async Task<T> WaitForPromiseAsync<T>(Task<T> task, Action dispose = nul
             Dispose();
             throw new TimeoutException(ex.Message + FormatLogRecording(_logs), ex);
         }
+        catch (TaskCanceledException ex)
+        {
+            dispose?.Invoke();
+            _error = ex.ToString();
+            Dispose();
+            return default;
```
Comment (campersau) [inline src/Playwright/Core/Waiter.cs:238]: To me it would be strange to just get the default value back instead of a canceled exception.
In .NET it is standard that the user needs to handle canceled exceptions as soon as they are passing cancellation tokens around.

File: src/Playwright/Core/Waiter.cs [hunk 3]
```diff
@@ -218,6 +230,13 @@ internal async Task<T> WaitForPromiseAsync<T>(Task<T> task, Action dispose = nul
             Dispose();
             throw new TimeoutException(ex.Message + FormatLogRecording(_logs), ex);
         }
+        catch (TaskCanceledException ex)
```
Comment (campersau) [inline src/Playwright/Core/Waiter.cs:233]: Catch `OperationCanceledException` instead or both.
See https://github.com/dotnet/runtime/issues/83406#issue-1623962011

File: src/Playwright/Core/Waiter.cs [hunk 4]
```diff
@@ -117,6 +117,18 @@ internal void RejectImmediately(Exception exception)
...
 
+    internal void RejectOnCancellation(CancellationToken cancellationToken)
+    {
+        if (!cancellationToken.CanBeCanceled)
+        {
+            return;
+        }
+
+        var cts = new TaskCompletionSource<bool>();
+        var registration = cancellationToken.Register(cts.SetCanceled, false);
```
Comment (campersau) [inline src/Playwright/Core/Waiter.cs:128]: If you let the cancellation throw instead (https://github.com/microsoft/playwright-dotnet/pull/2597#discussion_r1211064250) this should use the `Set Canceled(cancellationToken)` instead so users can inspect it if the catch it higher up in the stack.

File: src/Playwright/Core/Waiter.cs [hunk 5]
```diff
@@ -117,6 +117,18 @@ internal void RejectImmediately(Exception exception)
         _immediateError = exception;
     }
 
+    internal void RejectOnCancellation(CancellationToken cancellationToken)
+    {
+        if (!cancellationToken.CanBeCanceled)
+        {
+            return;
+        }
+
```
Comment (campersau) [inline src/Playwright/Core/Waiter.cs:126]: Just an idea: maybe it would be good to call `cancellationToken.ThrowIfCancellationRequested()` (here or maybe higher up in the call stack) so that if it is already cancelled we don't do anything.

File: src/Playwright/Core/Waiter.cs (hunk 2 above)
Comment (maw136) [reply src/Playwright/Core/Waiter.cs:238]: I was looking for an accepted general rule for that. Do you suggest having that specific exception bubble up or converting it (as by default) to PlaywrightException?

Comment (campersau) [reply src/Playwright/Core/Waiter.cs:238]: From my experience cancellation exceptions escpecially ones triggered by a user provided CancellationToken aren't wrapped / converted in other exceptions. So I would vote for letting the cancellation exception bubble up.

Comment (maw136) [reply src/Playwright/Core/Waiter.cs]: I've refactored into rethrow.

File: src/Playwright/Core/Waiter.cs (hunk 3 above)
Comment (maw136) [reply src/Playwright/Core/Waiter.cs]: Yeah, missed that inheritance hierarchy. Fixed

File: src/Playwright/Core/Waiter.cs (hunk 5 above)
Comment (maw136) [reply src/Playwright/Core/Waiter.cs:126]: Yes, that was my understanding of your comment as well, alas I have added the throw in the wrong place, fixed now :)

File: src/Playwright/Core/Waiter.cs [hunk 6]
```diff
@@ -243,7 +267,11 @@ private static async Task WrapActionAsync(Func<Task> action, CancellationTokenSo
         }
         catch
         {
-            cts.Cancel();
+            if (!cts.IsCancellationRequested)
```
Comment (mxschmitt) [inline src/Playwright/Core/Waiter.cs:270]: Why this change? What breaks if we always cancel here?

File: src/Playwright/Core/Waiter.cs [hunk 7]
```diff
@@ -218,6 +235,13 @@ internal async Task<T> WaitForPromiseAsync<T>(Task<T> task, Action dispose = nul
             Dispose();
             throw new TimeoutException(ex.Message + FormatLogRecording(_logs), ex);
         }
+        catch (OperationCanceledException ex)
+        {
+            dispose?.Invoke();
+            _error = ex.ToString();
+            Dispose();
+            throw;
```
Comment (mxschmitt) [inline src/Playwright/Core/Waiter.cs:243]: would be nice to throw the same exception but with formatted logs like in line 250.

File: src/Playwright/API/Generated/Options/BrowserContextRunAndWaitForConsoleMessageOptions.cs [hunk 8]
```diff
@@ -62,6 +64,14 @@ public BrowserContextRunAndWaitForConsoleMessageOptions(BrowserContextRunAndWait
...
     [JsonPropertyName("timeout")]
     public float? Timeout { get; set; }
+
+    /// <summary>
+    /// <para>
+    /// Cancellation token to provide cancellation of waiting for console message
+    /// </para>
+    /// </summary>
+    [JsonPropertyName("cancellationToken")]
+    public CancellationToken CancellationToken { get; set; }
```
Comment (mxschmitt) [inline src/Playwright/API/Generated/Options/BrowserContextRunAndWaitForConsoleMessageOptions.cs:74]: Did you add these manually? All these files in that folder are generated automatically. We should add csharp specific properties upstream [like here](https://github.com/search?q=repo%3Amicrosoft%2Fplaywright+RunAndWaitFor+language%3AMarkdown&type=code) so that the generator will pick it up.

Comment (maw136) [reply src/Playwright/API/Generated/Options/BrowserContextRunAndWaitForConsoleMessageOptions.cs:74]: Yes, manual addition. I will move it to the source for the generator.

Comment (mxschmitt) [reply src/Playwright/API/Generated/Options/BrowserContextRunAndWaitForConsoleMessageOptions.cs:74]: See [here](https://github.com/microsoft/playwright-dotnet/blob/main/CONTRIBUTING.md#generating-the-apis--rolling-the-driver-upstream) for some notes on how it works.

File: src/Playwright/Core/Waiter.cs (hunk 6 above)
Comment (maw136) [reply src/Playwright/Core/Waiter.cs:270]: There is no need to cancel 'harder'. I have replicated the solution from the other place in this PR where this check is made, and I remembered that I have added this cancellation here some time ago. This is the correct solution for the overall problem. Any code should be written in the most proper/correct way. Simply working is not ever good enough.

File: src/Playwright/Core/Waiter.cs (hunk 7 above)
Comment (maw136) [reply src/Playwright/Core/Waiter.cs:243]: With `OperationCanceledException` the suggested flow is to throw the original exception to have better understanding of the place of cancelation. https://devblogs.microsoft.com/premier-developer/recommended-patterns-for-cancellationtoken/

**Comment by maw136** [comment]
Added PR for upstream:
https://github.com/microsoft/playwright/pull/23953

File: src/Playwright/API/Generated/Options/BrowserContextRunAndWaitForConsoleMessageOptions.cs (hunk 8 above)
Comment (maw136) [reply src/Playwright/API/Generated/Options/BrowserContextRunAndWaitForConsoleMessageOptions.cs:74]: https://github.com/microsoft/playwright/pull/23953

**Comment by maw136** [comment]
Used this generator for generated members: 
https://github.com/microsoft/playwright/pull/23953

**Comment by mxschmitt** [comment]
Closing as per https://github.com/microsoft/playwright/issues/24083#issuecomment-1640761745.
